### DISCUSSION TITLE
Bug in grep

### DISCUSSION THREAD

**Discussion by vschulz** [git-for-windows/git#2892] [Q&A] [answered]
- [X ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?
64-bit
```
$ git --version --build-options

git version 2.29.2.windows.2
cpu: x86_64
built from commit: 3464b98ce6803c98bf8fb34390cd150d66e4a0d3
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?
10 64-bit
```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19041.572]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: Notepad++
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

** insert your response here **

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

** insert your response here **
CMD
 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
C:\Program Files\Git\usr\bin>grep.exe '^20'  C:\Users\vpsch\Desktop\test.txt

C:\Program Files\Git\usr\bin>cat.exe   C:\Users\vpsch\Desktop\test.txt
```
 - What did you expect to occur after running these commands?

20 should be found

20 should be found
x20 should not be found
 - What actually happened instead?

20 should be found
x20 should not be found

20 should be found
x20 should not be found

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

** insert URL here **

**Comment by dscho** [comment]
>     * What commands did you run to trigger this issue? If you can provide a
>       [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
>       this will help us understand the issue.
> 
> 
> ```
> C:\Program Files\Git\usr\bin>grep.exe '^20'  C:\Users\vpsch\Desktop\test.txt
> 
> C:\Program Files\Git\usr\bin>cat.exe   C:\Users\vpsch\Desktop\test.txt
> ```
> 
>     * What did you expect to occur after running these commands?
> 
> 
> 20 should be found
> 
> 20 should be found
> x20 should not be found
> 
>     * What actually happened instead?
> 
> 
> 20 should be found
> x20 should not be found
> 
> 20 should be found
> x20 should not be found

I am embarrassed to admit that I do not understand what this report is trying to tell me. Did you mean to show the actual contents of `test.txt` and the actual `grep` output?

**Comment by vschulz** [comment]
Sorry for the lack of clarity.  The contents of test.txt is 
20 should be found
x20 should not be found

grep is ignoring the ^ in pattern matching, so grep ^20 is matching in line with text  x20 should not be found.  Grep works properly in ubuntu 20 command line etc, and also in the versions in C:\Program Files\Git\bin\bash.exe and sh.exe.  This affects other software that imports git-for-windows such as cmder

**Comment by Project-Magenta** [comment, marked as answer]
Are you sure there's a bug? ^ is the shell escape character in cmd.exe

**Comment by landstander668** [comment]
As Project-Magenta mentioned, I'm pretty sure this is due to `cmd.exe` treating ^ as an escape character. Here's a quick test to demonstrate its effect:

```
C:\Users\adric> echo '^20'
'20'

C:\Users\adric> echo '^^20'
'^20'
```

**Comment by landstander668** [comment]
>
> so kind of like a Git CMD bug
>

Not really a bug per se... more of a difference in expectation vs
implementation. This was an artifact of Windows' built-in cmd.exe
interpreter, which simply behaves differently than a unix-like shell...
it's handling of quoted arguments being another notable example, IIRC. As a
result, it's not really all that uncommon to need to tweak command syntax
slightly when working on the Git CMD side of the fence. The differences can
most certainly be confusing, however.

-- 
"In the beginning the Universe was created. This has made a lot of people
very angry and been widely regarded as a bad move." -Douglas Adams

**Comment by Project-Magenta** [comment]
that's why I edited
